File: android/src/main/kotlin/re/notifica/flutter/handlers/DeviceManagerHandler.kt
package re.notifica.flutter.handlers

import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import re.notifica.Notificare
import re.notifica.NotificareCallback
import re.notifica.callbacks.*
import re.notifica.flutter.NotificarePlugin
import re.notifica.flutter.models.NotificareDoNotDisturb
import re.notifica.flutter.models.toFlutterMap
import re.notifica.flutter.utils.onMainThread
import re.notifica.models.NotificareDoNotDisturb
import re.notifica.models.NotificareUserData

class DeviceManagerHandler : MethodChannel.MethodCallHandler {

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "getCurrentDevice" -> getCurrentDevice(result)
            "register" -> register(call, result)
            "fetchTags" -> fetchTags(result)
            "addTag" -> addTag(call, result)
            "addTags" -> addTags(call, result)
            "removeTag" -> removeTag(call, result)
            "removeTags" -> removeTags(call, result)
            "clearTags" -> clearTags(result)
            "getPreferredLanguage" -> getPreferredLanguage(result)
            "updatePreferredLanguage" -> updatePreferredLanguage(call, result)
            "fetchDoNotDisturb" -> fetchDoNotDisturb(result)
            "updateDoNotDisturb" -> updateDoNotDisturb(call, result)
            "clearDoNotDisturb" -> clearDoNotDisturb(result)
            "fetchUserData" -> fetchUserData(result)
            "updateUserData" -> updateUserData(call, result)
            else -> result.notImplemented()
        }
    }

    private fun getCurrentDevice(result: MethodChannel.Result) {
        result.success(Notificare.deviceManager.currentDevice?.toFlutterMap())
    }

    private fun register(call: MethodCall, methodResult: MethodChannel.Result) {
        val userId = call.argument<String?>("userId")
        val userName = call.argument<String?>("userName")

        Notificare.deviceManager.register(userId, userName, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                methodResult.onMainThread().success(null)
            }

            override fun onFailure(e: Exception) {
                methodResult.onMainThread().error(NotificarePlugin.DEFAULT_ERROR_CODE, e.message, null)
            }
        })
    }

    private fun fetchTags(methodResult: MethodChannel.Result) {
        Notificare.deviceManager.fetchTags(object : NotificareCallback<List<String>> {
            override fun onSuccess(result: List<String>) {
                methodResult.onMainThread().success(result)
            }

            override fun onFailure(e: Exception) {
                methodResult.onMainThread().error(NotificarePlugin.DEFAULT_ERROR_CODE, e.message, null)
            }
        })
    }

    private fun addTag(call: MethodCall, methodResult: MethodChannel.Result) {
        val tag = call.arguments<String>()

        Notificare.deviceManager.addTag(tag, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                methodResult.onMainThread().success(null)
            }

            override fun onFailure(e: Exception) {
                methodResult.onMainThread().error(NotificarePlugin.DEFAULT_ERROR_CODE, e.message, null)
            }
        })
    }

    private fun addTags(call: MethodCall, methodResult: MethodChannel.Result) {
        val tags = call.arguments<List<String>>()

        Notificare.deviceManager.addTags(tags, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                methodResult.onMainThread().success(null)
            }

            override fun onFailure(e: Exception) {
                methodResult.onMainThread().error(NotificarePlugin.DEFAULT_ERROR_CODE, e.message, null)
            }
        })
    }

    private fun removeTag(call: MethodCall, methodResult: MethodChannel.Result) {
        val tag = call.arguments<String>()

        Notificare.deviceManager.removeTag(tag, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                methodResult.onMainThread().success(null)
            }

            override fun onFailure(e: Exception) {
                methodResult.onMainThread().error(NotificarePlugin.DEFAULT_ERROR_CODE, e.message, null)
            }
        })
    }

    private fun removeTags(call: MethodCall, methodResult: MethodChannel.Result) {
        val tags = call.arguments<List<String>>()

        Notificare.deviceManager.removeTags(tags, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                methodResult.onMainThread().success(null)
            }

            override fun onFailure(e: Exception) {
                methodResult.onMainThread().error(NotificarePlugin.DEFAULT_ERROR_CODE, e.message, null)
            }
        })
    }

    private fun clearTags(methodResult: MethodChannel.Result) {
        Notificare.deviceManager.clearTags(object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                methodResult.onMainThread().success(null)
            }

            override fun onFailure(e: Exception) {
                methodResult.onMainThread().error(NotificarePlugin.DEFAULT_ERROR_CODE, e.message, null)
            }
        })
    }

    private fun getPreferredLanguage(result: MethodChannel.Result) {
        result.success(Notificare.deviceManager.preferredLanguage)
    }

    private fun updatePreferredLanguage(call: MethodCall, methodResult: MethodChannel.Result) {
        val language = call.arguments<String?>()

        Notificare.deviceManager.updatePreferredLanguage(language, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                methodResult.onMainThread().success(null)
            }

            override fun onFailure(e: Exception) {
                methodResult.onMainThread().error(NotificarePlugin.DEFAULT_ERROR_CODE, e.message, null)
            }
        })
    }

    private fun fetchDoNotDisturb(methodResult: MethodChannel.Result) {
        Notificare.deviceManager.fetchDoNotDisturb(object : NotificareCallback<NotificareDoNotDisturb?> {
            override fun onSuccess(result: NotificareDoNotDisturb?) {
                methodResult.onMainThread().success(result?.toFlutterMap())
            }

            override fun onFailure(e: Exception) {
                methodResult.onMainThread().error(NotificarePlugin.DEFAULT_ERROR_CODE, e.message, null)
            }
        })
    }

    private fun updateDoNotDisturb(call: MethodCall, methodResult: MethodChannel.Result) {
        val dnd = NotificareDoNotDisturb(call.arguments())

        Notificare.deviceManager.updateDoNotDisturb(dnd, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                methodResult.onMainThread().success(null)
            }

            override fun onFailure(e: Exception) {
                methodResult.onMainThread().error(NotificarePlugin.DEFAULT_ERROR_CODE, e.message, null)
            }
        })
    }

    private fun clearDoNotDisturb(methodResult: MethodChannel.Result) {
        Notificare.deviceManager.clearDoNotDisturb(object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                methodResult.onMainThread().success(null)
            }

            override fun onFailure(e: Exception) {
                methodResult.onMainThread().error(NotificarePlugin.DEFAULT_ERROR_CODE, e.message, null)
            }
        })
    }

    private fun fetchUserData(methodResult: MethodChannel.Result) {
        Notificare.deviceManager.fetchUserData(object : NotificareCallback<NotificareUserData?> {
            override fun onSuccess(result: NotificareUserData?) {
                methodResult.onMainThread().success(result)
            }

            override fun onFailure(e: Exception) {
                methodResult.onMainThread().error(NotificarePlugin.DEFAULT_ERROR_CODE, e.message, null)
            }
        })
    }

    private fun updateUserData(call: MethodCall, methodResult: MethodChannel.Result) {
        val userData = call.arguments<Map<String, String>>()

        Notificare.deviceManager.updateUserData(userData, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                methodResult.onMainThread().success(null)
            }

            override fun onFailure(e: Exception) {
                methodResult.onMainThread().error(NotificarePlugin.DEFAULT_ERROR_CODE, e.message, null)
            }
        })
    }
}

File: ios/Classes/Handlers/DeviceManagerHandler.swift
//
//  DeviceManagerHandler.swift
//  notificare
//
//  Created by Helder Pinhal on 01/12/2020.
//

import Foundation
import NotificareSDK

class DeviceManagerHandler {

    func handler(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getCurrentDevice": getCurrentDevice(call, result)
        case "register": register(call, result)
        case "fetchTags": fetchTags(call, result)
        case "addTag": addTag(call, result)
        case "addTags": addTags(call, result)
        case "removeTag": removeTag(call, result)
        case "removeTags": removeTags(call, result)
        case "clearTags": clearTags(call, result)
        case "getPreferredLanguage": getPreferredLanguage(call, result)
        case "updatePreferredLanguage": updatePreferredLanguage(call, result)
        case "fetchDoNotDisturb": fetchDoNotDisturb(call, result)
        case "updateDoNotDisturb": updateDoNotDisturb(call, result)
        case "clearDoNotDisturb": clearDoNotDisturb(call, result)
        case "fetchUserData": fetchUserData(call, result)
        case "updateUserData": updateUserData(call, result)
        default: result(FlutterMethodNotImplemented)
        }
    }

    private func getCurrentDevice(_ call: FlutterMethodCall, _ response: @escaping FlutterResult) {
        response(Notificare.shared.deviceManager.currentDevice?.toDictionary())
    }

    private func register(_ call: FlutterMethodCall, _ response: @escaping FlutterResult) {
        let arguments = call.arguments as! FlutterDictionary

        Notificare.shared.deviceManager.register(
            userId: arguments["userId"] as? String,
            userName: arguments["userName"] as? String
        ) { result in
            switch result {
            case .success:
                response(nil)
            case .failure(let error):
                response(FlutterError(code: DEFAULT_ERROR_CODE, message: error.localizedDescription, details: nil))
            }
        }
    }

    private func fetchTags(_ call: FlutterMethodCall, _ response: @escaping FlutterResult) {
        Notificare.shared.deviceManager.fetchTags { result in
            switch result {
            case .success(let tags):
                response(tags)
            case .failure(let error):
                response(FlutterError(code: DEFAULT_ERROR_CODE, message: error.localizedDescription, details: nil))
            }
        }
    }

    private func addTag(_ call: FlutterMethodCall, _ response: @escaping FlutterResult) {
        let tag = call.arguments as! String

        Notificare.shared.deviceManager.addTag(tag) { result in
            switch result {
            case .success:
                response(nil)
            case .failure(let error):
                response(FlutterError(code: DEFAULT_ERROR_CODE, message: error.localizedDescription, details: nil))
            }
        }
    }

    private func addTags(_ call: FlutterMethodCall, _ response: @escaping  FlutterResult) {
        let tags = call.arguments as! [String]

        Notificare.shared.deviceManager.addTags(tags) { result in
            switch result {
            case .success:
                response(nil)
            case .failure(let error):
                response(FlutterError(code: DEFAULT_ERROR_CODE, message: error.localizedDescription, details: nil))
            }
        }
    }

    private func removeTag(_ call: FlutterMethodCall, _ response: @escaping  FlutterResult) {
        let tag = call.arguments as! String

        Notificare.shared.deviceManager.removeTag(tag) { result in
            switch result {
            case .success:
                response(nil)
            case .failure(let error):
                response(FlutterError(code: DEFAULT_ERROR_CODE, message: error.localizedDescription, details: nil))
            }
        }
    }

    private func removeTags(_ call: FlutterMethodCall, _ response: @escaping  FlutterResult) {
        let tags = call.arguments as! [String]

        Notificare.shared.deviceManager.removeTags(tags) { result in
            switch result {
            case .success:
                response(nil)
            case .failure(let error):
                response(FlutterError(code: DEFAULT_ERROR_CODE, message: error.localizedDescription, details: nil))
            }
        }
    }

    private func clearTags(_ call: FlutterMethodCall, _ response: @escaping  FlutterResult) {
        Notificare.shared.deviceManager.clearTags { result in
            switch result {
            case .success:
                response(nil)
            case .failure(let error):
                response(FlutterError(code: DEFAULT_ERROR_CODE, message: error.localizedDescription, details: nil))
            }
        }
    }

    private func getPreferredLanguage(_ call: FlutterMethodCall, _ response: @escaping  FlutterResult) {
        response(Notificare.shared.deviceManager.preferredLanguage)
    }

    private func updatePreferredLanguage(_ call: FlutterMethodCall, _ response: @escaping  FlutterResult) {
        let language = call.arguments as! String?

        Notificare.shared.deviceManager.updatePreferredLanguage(language) { result in
            switch result {
            case .success:
                response(nil)
            case .failure(let error):
                response(FlutterError(code: DEFAULT_ERROR_CODE, message: error.localizedDescription, details: nil))
            }
        }
    }

    private func fetchDoNotDisturb(_ call: FlutterMethodCall, _ response: @escaping  FlutterResult) {
        Notificare.shared.deviceManager.fetchDoNotDisturb { result in
            switch result {
            case .success(let dnd):
                response(dnd?.toDictionary())
            case .failure(let error):
                response(FlutterError(code: DEFAULT_ERROR_CODE, message: error.localizedDescription, details: nil))
            }
        }
    }

    private func updateDoNotDisturb(_ call: FlutterMethodCall, _ response: @escaping  FlutterResult) {
        let arguments = call.arguments as! FlutterDictionary
        let dnd: NotificareDoNotDisturb

        do {
            dnd = try NotificareDoNotDisturb(dictionary: arguments)
        } catch {
            response(FlutterError(code: DEFAULT_ERROR_CODE, message: error.localizedDescription, details: nil))
            return
        }

        Notificare.shared.deviceManager.updateDoNotDisturb(dnd) { result in
            switch result {
            case .success:
                response(nil)
            case .failure(let error):
                response(FlutterError(code: DEFAULT_ERROR_CODE, message: error.localizedDescription, details: nil))
            }
        }
    }

    private func clearDoNotDisturb(_ call: FlutterMethodCall, _ response: @escaping  FlutterResult) {
        Notificare.shared.deviceManager.clearDoNotDisturb { result in
            switch result {
            case .success:
                response(nil)
            case .failure(let error):
                response(FlutterError(code: DEFAULT_ERROR_CODE, message: error.localizedDescription, details: nil))
            }
        }
    }

    private func fetchUserData(_ call: FlutterMethodCall, _ response: @escaping  FlutterResult) {
        Notificare.shared.deviceManager.fetchUserData { result in
            switch result {
            case .success(let userData):
                response(userData)
            case .failure(let error):
                response(FlutterError(code: DEFAULT_ERROR_CODE, message: error.localizedDescription, details: nil))
            }
        }
    }

    private func updateUserData(_ call: FlutterMethodCall, _ response: @escaping  FlutterResult) {
        let userData = call.arguments as! [String: String]

        Notificare.shared.deviceManager.updateUserData(userData) { result in
            switch result {
            case .success:
                response(nil)
            case .failure(let error):
                response(FlutterError(code: DEFAULT_ERROR_CODE, message: error.localizedDescription, details: nil))
            }
        }
    }
}
